File: framework/validator.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Tuple

from .models import Assertion, Scenario, Step


ALLOWED_TOP = {"meta", "config", "steps", "assertions", "output"}
ALLOWED_META = {"name", "version", "platform", "tags"}
ALLOWED_CONFIG = {"timeoutMs", "seed", "artifactsDir"}
ALLOWED_STEP = {"id", "command", "args"}
ALLOWED_ASSERT = {"id", "type", "target"}
ALLOWED_OUTPUT = {"report"}
SUPPORTED_COMMANDS = {"tap", "navigate", "act", "screenshot", "wait", "source"}


def _validate_tap_args(args: Dict[str, Any], path: str) -> List[ValidationIssue]:
    issues: List[ValidationIssue] = []
    has_target = "target" in args
    has_x = "x" in args
    has_y = "y" in args

    if has_target:
        if has_x or has_y:
            issues.append(ValidationIssue("error", "ARG_ERROR", "tap cannot mix args.target with args.x/args.y", path))
        return issues

    if has_x != has_y:
        issues.append(ValidationIssue("error", "ARG_ERROR", "tap coordinate mode requires both args.x and args.y", path))
        return issues

    if not has_x and not has_y:
        issues.append(ValidationIssue("error", "ARG_ERROR", "tap requires args.target or args.x/args.y", path))
        return issues

    if "normalized" in args and not isinstance(args["normalized"], bool):
        issues.append(ValidationIssue("error", "ARG_ERROR", "tap args.normalized must be boolean", path))
    return issues


@dataclass
class ValidationIssue:
    severity: str  # error | warning
    code: str
    message: str
    path: str


def _unknown_fields(obj: Dict[str, Any], allowed: set[str], path: str) -> List[ValidationIssue]:
    issues: List[ValidationIssue] = []
    for key in obj.keys():
        if key not in allowed:
            issues.append(
                ValidationIssue(
                    severity="error",
                    code="UNKNOWN_FIELD",
                    message=f"Unknown field '{key}'",
                    path=f"{path}.{key}" if path else key,
                )
            )
    return issues


def _required(obj: Dict[str, Any], fields: List[str], path: str) -> List[ValidationIssue]:
    issues: List[ValidationIssue] = []
    for f in fields:
        if f not in obj:
            issues.append(ValidationIssue("error", "MISSING_REQUIRED", f"Missing required field '{f}'", path))
    return issues


def parse_and_validate(path: str) -> Tuple[Scenario | None, List[ValidationIssue]]:
    issues: List[ValidationIssue] = []
    raw = json.loads(Path(path).read_text(encoding="utf-8"))

    if not isinstance(raw, dict):
        return None, [ValidationIssue("error", "TYPE_ERROR", "Scenario root must be object", "$")]

    issues.extend(_unknown_fields(raw, ALLOWED_TOP, "$"))
    issues.extend(_required(raw, ["meta", "config", "steps"], "$"))

    meta = raw.get("meta", {})
    config = raw.get("config", {})
    steps_raw = raw.get("steps", [])
    asserts_raw = raw.get("assertions", [])
    output = raw.get("output", {})

    if isinstance(meta, dict):
        issues.extend(_unknown_fields(meta, ALLOWED_META, "$.meta"))
        issues.extend(_required(meta, ["name", "version", "platform"], "$.meta"))
    else:
        issues.append(ValidationIssue("error", "TYPE_ERROR", "meta must be object", "$.meta"))

    if isinstance(config, dict):
        issues.extend(_unknown_fields(config, ALLOWED_CONFIG, "$.config"))
    else:
        issues.append(ValidationIssue("error", "TYPE_ERROR", "config must be object", "$.config"))

    steps: List[Step] = []
    if not isinstance(steps_raw, list) or not steps_raw:
        issues.append(ValidationIssue("error", "TYPE_ERROR", "steps must be non-empty list", "$.steps"))
    else:
        for i, s in enumerate(steps_raw):
            p = f"$.steps[{i}]"
            if not isinstance(s, dict):
                issues.append(ValidationIssue("error", "TYPE_ERROR", "step must be object", p))
                continue
            issues.extend(_unknown_fields(s, ALLOWED_STEP, p))
            issues.extend(_required(s, ["id", "command", "args"], p))
            cmd = s.get("command")
            if isinstance(cmd, str) and cmd not in SUPPORTED_COMMANDS:
                issues.append(ValidationIssue("error", "UNSUPPORTED_COMMAND", f"Unsupported command '{cmd}'", f"{p}.command"))
            if cmd == "tap" and isinstance(s.get("args"), dict):
                issues.extend(_validate_tap_args(s["args"], f"{p}.args"))
            if cmd == "navigate" and isinstance(s.get("args"), dict) and "to" not in s["args"]:
                issues.append(ValidationIssue("error", "ARG_ERROR", "navigate requires args.to", f"{p}.args"))
            if cmd == "screenshot" and isinstance(s.get("args"), dict) and "label" not in s["args"]:
                issues.append(ValidationIssue("warning", "DETERMINISM_WARNING", "screenshot missing label may reduce determinism", f"{p}.args"))
            if cmd == "wait" and isinstance(s.get("args"), dict) and "ms" not in s["args"]:
                issues.append(ValidationIssue("error", "ARG_ERROR", "wait requires args.ms", f"{p}.args"))
            if isinstance(s.get("id"), str) and isinstance(cmd, str) and isinstance(s.get("args"), dict):
                steps.append(Step(id=s["id"], command=cmd, args=s["args"]))

    assertions: List[Assertion] = []
    if asserts_raw is None:
        asserts_raw = []
    if isinstance(asserts_raw, list):
        for i, a in enumerate(asserts_raw):
            p = f"$.assertions[{i}]"
            if not isinstance(a, dict):
                issues.append(ValidationIssue("error", "TYPE_ERROR", "assertion must be object", p))
                continue
            issues.extend(_unknown_fields(a, ALLOWED_ASSERT, p))
            issues.extend(_required(a, ["id", "type", "target"], p))
            if all(k in a for k in ["id", "type", "target"]):
                assertions.append(Assertion(id=a["id"], type=a["type"], target=a["target"]))
    else:
        issues.append(ValidationIssue("error", "TYPE_ERROR", "assertions must be list", "$.assertions"))

    if isinstance(output, dict):
        issues.extend(_unknown_fields(output, ALLOWED_OUTPUT, "$.output"))
    else:
        issues.append(ValidationIssue("error", "TYPE_ERROR", "output must be object", "$.output"))

    platform = meta.get("platform") if isinstance(meta, dict) else None
    if platform not in {"ios", "android"}:
        issues.append(ValidationIssue("error", "INPUT_ERROR", "meta.platform must be ios|android", "$.meta.platform"))

    if isinstance(config, dict) and "seed" not in config:
        issues.append(ValidationIssue("warning", "DETERMINISM_WARNING", "config.seed is missing", "$.config"))

    if any(i.severity == "error" for i in issues):
        return None, issues

    scenario = Scenario(
        meta=meta,
        config=config,
        steps=steps,
        assertions=assertions,
        output=output,
    )
    return scenario, issues

File: src/validator.ts
import fs from 'node:fs';

import type {
  Assertion,
  ParseValidationResult,
  Platform,
  Scenario,
  Step,
  ValidationIssue
} from './types.js';

const ALLOWED_TOP = new Set(['meta', 'config', 'steps', 'assertions', 'output']);
const ALLOWED_META = new Set(['name', 'version', 'platform', 'tags']);
const ALLOWED_CONFIG = new Set(['timeoutMs', 'seed', 'artifactsDir']);
const ALLOWED_STEP = new Set(['id', 'command', 'args']);
const ALLOWED_ASSERT = new Set(['id', 'type', 'target']);
const ALLOWED_OUTPUT = new Set(['report']);
const SUPPORTED_COMMANDS = new Set(['tap', 'navigate', 'act', 'scroll', 'screenshot', 'wait', 'source']);

function validationIssue(
  severity: 'error' | 'warning',
  code: string,
  message: string,
  issuePath: string
): ValidationIssue {
  return { severity, code, message, path: issuePath };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

function unknownFields(
  obj: Record<string, unknown>,
  allowed: Set<string>,
  issuePath: string
): ValidationIssue[] {
  return Object.keys(obj)
    .filter((key) => !allowed.has(key))
    .map((key) =>
      validationIssue(
        'error',
        'UNKNOWN_FIELD',
        `Unknown field '${key}'`,
        issuePath ? `${issuePath}.${key}` : key
      )
    );
}

function requiredFields(
  obj: Record<string, unknown>,
  fields: string[],
  issuePath: string
): ValidationIssue[] {
  return fields
    .filter((field) => !(field in obj))
    .map((field) =>
      validationIssue('error', 'MISSING_REQUIRED', `Missing required field '${field}'`, issuePath)
    );
}

function validateTapArgs(args: Record<string, unknown>, issuePath: string): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  const hasTarget = Object.hasOwn(args, 'target');
  const hasX = Object.hasOwn(args, 'x');
  const hasY = Object.hasOwn(args, 'y');

  if (hasTarget) {
    if (hasX || hasY) {
      issues.push(
        validationIssue(
          'error',
          'ARG_ERROR',
          'tap cannot mix args.target with args.x/args.y',
          issuePath
        )
      );
    }
    return issues;
  }

  if (hasX !== hasY) {
    issues.push(
      validationIssue(
        'error',
        'ARG_ERROR',
        'tap coordinate mode requires both args.x and args.y',
        issuePath
      )
    );
    return issues;
  }

  if (!hasX && !hasY) {
    issues.push(
      validationIssue(
        'error',
        'ARG_ERROR',
        'tap requires args.target or args.x/args.y',
        issuePath
      )
    );
    return issues;
  }

  if (Object.hasOwn(args, 'normalized') && typeof args.normalized !== 'boolean') {
    issues.push(
      validationIssue('error', 'ARG_ERROR', 'tap args.normalized must be boolean', issuePath)
    );
  }

  return issues;
}

function validateScrollArgs(args: Record<string, unknown>, issuePath: string): ValidationIssue[] {
  const issues: ValidationIssue[] = [];

  if (!Object.hasOwn(args, 'direction')) {
    issues.push(
      validationIssue('error', 'ARG_ERROR', 'scroll requires args.direction', issuePath)
    );
    return issues;
  }

  if (typeof args.direction !== 'string' || !['up', 'down'].includes(args.direction.toLowerCase())) {
    issues.push(
      validationIssue(
        'error',
        'ARG_ERROR',
        "scroll args.direction must be 'up' or 'down'",
        issuePath
      )
    );
  }

  if (
    Object.hasOwn(args, 'percent') &&
    (typeof args.percent !== 'number' || !Number.isFinite(args.percent) || args.percent < 1 || args.percent > 100)
  ) {
    issues.push(
      validationIssue(
        'error',
        'ARG_ERROR',
        'scroll args.percent must be a number between 1 and 100',
        issuePath
      )
    );
  }

  return issues;
}

export function parseAndValidate(filePath: string): ParseValidationResult {
  const issues: ValidationIssue[] = [];
  const raw = JSON.parse(fs.readFileSync(filePath, 'utf8')) as unknown;

  if (!isRecord(raw)) {
    return {
      scenario: null,
      issues: [validationIssue('error', 'TYPE_ERROR', 'Scenario root must be object', '$')]
    };
  }

  issues.push(...unknownFields(raw, ALLOWED_TOP, '$'));
  issues.push(...requiredFields(raw, ['meta', 'config', 'steps'], '$'));

  const metaRaw = raw.meta;
  const configRaw = raw.config;
  const stepsRaw = raw.steps;
  const assertionsRaw = raw.assertions ?? [];
  const outputRaw = raw.output ?? {};

  let meta: Scenario['meta'] | null = null;
  if (isRecord(metaRaw)) {
    issues.push(...unknownFields(metaRaw, ALLOWED_META, '$.meta'));
    issues.push(...requiredFields(metaRaw, ['name', 'version', 'platform'], '$.meta'));

    if (metaRaw.platform !== 'ios' && metaRaw.platform !== 'android') {
      issues.push(
        validationIssue('error', 'INPUT_ERROR', 'meta.platform must be ios|android', '$.meta.platform')
      );
    } else if (typeof metaRaw.name === 'string' && typeof metaRaw.version === 'string') {
      meta = {
        ...metaRaw,
        name: metaRaw.name,
        version: metaRaw.version,
        platform: metaRaw.platform as Platform
      };
    }
  } else {
    issues.push(validationIssue('error', 'TYPE_ERROR', 'meta must be object', '$.meta'));
  }

  const config = isRecord(configRaw) ? configRaw : {};
  if (isRecord(configRaw)) {
    issues.push(...unknownFields(configRaw, ALLOWED_CONFIG, '$.config'));
  } else {
    issues.push(validationIssue('error', 'TYPE_ERROR', 'config must be object', '$.config'));
  }

  if (!Object.hasOwn(config, 'seed')) {
    issues.push(validationIssue('warning', 'DETERMINISM_WARNING', 'config.seed is missing', '$.config'));
  }

  const steps: Step[] = [];
  if (!Array.isArray(stepsRaw) || stepsRaw.length === 0) {
    issues.push(validationIssue('error', 'TYPE_ERROR', 'steps must be non-empty list', '$.steps'));
  } else {
    for (const [index, rawStep] of stepsRaw.entries()) {
      const issuePath = `$.steps[${index}]`;
      if (!isRecord(rawStep)) {
        issues.push(validationIssue('error', 'TYPE_ERROR', 'step must be object', issuePath));
        continue;
      }

      issues.push(...unknownFields(rawStep, ALLOWED_STEP, issuePath));
      issues.push(...requiredFields(rawStep, ['id', 'command', 'args'], issuePath));

      const command = rawStep.command;
      const args = rawStep.args;
      if (typeof command === 'string' && !SUPPORTED_COMMANDS.has(command)) {
        issues.push(
          validationIssue(
            'error',
            'UNSUPPORTED_COMMAND',
            `Unsupported command '${command}'`,
            `${issuePath}.command`
          )
        );
      }

      if (command === 'tap' && isRecord(args)) {
        issues.push(...validateTapArgs(args, `${issuePath}.args`));
      }

      if (command === 'navigate' && isRecord(args) && !Object.hasOwn(args, 'to')) {
        issues.push(validationIssue('error', 'ARG_ERROR', 'navigate requires args.to', `${issuePath}.args`));
      }

      if (command === 'scroll' && isRecord(args)) {
        issues.push(...validateScrollArgs(args, `${issuePath}.args`));
      }

      if (command === 'screenshot' && isRecord(args) && !Object.hasOwn(args, 'label')) {
        issues.push(
          validationIssue(
            'warning',
            'DETERMINISM_WARNING',
            'screenshot missing label may reduce determinism',
            `${issuePath}.args`
          )
        );
      }

      if (command === 'wait' && isRecord(args) && !Object.hasOwn(args, 'ms')) {
        issues.push(validationIssue('error', 'ARG_ERROR', 'wait requires args.ms', `${issuePath}.args`));
      }

      if (
        typeof rawStep.id === 'string' &&
        typeof command === 'string' &&
        SUPPORTED_COMMANDS.has(command) &&
        isRecord(args)
      ) {
        steps.push({
          id: rawStep.id,
          command: command as Step['command'],
          args
        });
      }
    }
  }

  const assertions: Assertion[] = [];
  if (Array.isArray(assertionsRaw)) {
    for (const [index, rawAssertion] of assertionsRaw.entries()) {
      const issuePath = `$.assertions[${index}]`;
      if (!isRecord(rawAssertion)) {
        issues.push(validationIssue('error', 'TYPE_ERROR', 'assertion must be object', issuePath));
        continue;
      }

      issues.push(...unknownFields(rawAssertion, ALLOWED_ASSERT, issuePath));
      issues.push(...requiredFields(rawAssertion, ['id', 'type', 'target'], issuePath));

      if (
        typeof rawAssertion.id === 'string' &&
        typeof rawAssertion.type === 'string' &&
        typeof rawAssertion.target === 'string'
      ) {
        assertions.push({
          id: rawAssertion.id,
          type: rawAssertion.type,
          target: rawAssertion.target
        });
      }
    }
  } else {
    issues.push(validationIssue('error', 'TYPE_ERROR', 'assertions must be list', '$.assertions'));
  }

  const output = isRecord(outputRaw) ? outputRaw : {};
  if (isRecord(outputRaw)) {
    issues.push(...unknownFields(outputRaw, ALLOWED_OUTPUT, '$.output'));
  } else {
    issues.push(validationIssue('error', 'TYPE_ERROR', 'output must be object', '$.output'));
  }

  if (issues.some((issue) => issue.severity === 'error') || meta === null) {
    return {
      scenario: null,
      issues
    };
  }

  return {
    scenario: {
      meta,
      config,
      steps,
      assertions,
      output
    },
    issues
  };
}
